lib/fastlane/plugin/wpmreleasetoolkit/actions/ios/ios_generate_strings_file_from_code.rb

```diff
@@ -0,0 +1,74 @@
+require 'open3'
+
+module Fastlane
+  module Actions
+    class IosGenerateStringsFileFromCode < Action
+      def self.run(params)
+        globbed_paths = params[:paths].map { |p| File.file?(p) ? p : "#{p}/**/*.{m,swift}" }
+        files = Dir.glob(globbed_paths)
+
+        flags = [('-q' if params[:quiet]), ('-SwiftUI' if params[:swiftui])].compact
+
+        out, status = Open3.capture2e('genstrings', '-o', params[:output_dir], *flags, *files)
```

Comment:
This LGTM, is simple enough, and code that is easy to follow. Well done.

How do you feel about having the core of the action logic, which in this case would be the lines I highlighted, extracted in a dedicated object? 

I'm more interested in the approach in general, as it's possibly overkill here. When I worked on the `configure_*` actions, I wished the logic was extracted because it does so much in a single `run` call and it was hard to test in isolation.

Or, I guess we can just be happy with higher-abstraction, functional-like tests like the ones from this PR 🤔

Reply:
Another reason I'm wary of testing `Action` subclasses directly is that `UI.` calls can leak into the tests results, unless we explicitly account for it in the spec, e.g. with `allow(UI).to receive...`

```
➜ be rspec spec/ios_generate_strings_file_from_code_spec.rb

Fastlane::Actions::IosGenerateStringsFileFromCode
  when including pods
    Generates the expected .strings files with SwiftUI support
    Generates the expected .strings files without SwiftUI support
  when not including pods
    Generates the expected .strings files with SwiftUI support
    Generates the expected .strings files without SwiftUI support
  when genstrings finds warnings
    only logs warnings about multiple values in quiet mode
Logging disabled while running tests. Force them by setting the DEBUG environment variable 👈👈👈👈👈
    logs warnings about both duplicate values and duplicate comments if not in quiet mode
```

```
➜ DEBUG=1 be rspec spec/ios_generate_strings_file_from_code_spec.rb

Fastlane::Actions::IosGenerateStringsFileFromCode
  when including pods
    Generates the expected .strings files with SwiftUI support
    Generates the expected .strings files without SwiftUI support
  when not including pods
    Generates the expected .strings files with SwiftUI support
    Generates the expected .strings files without SwiftUI support
  when genstrings finds warnings
    only logs warnings about multiple values in quiet mode
👇👇👇👇
[15:24:16]: ▸ Key "app.key5" used with multiple values. Value "app value 5\nwith multiple lines." kept. Value "app value 5\nwith multiple lines, and different value than in Swift" ignored.
[15:24:16]: ▸ genstrings: warning: Key "app.key5" used with multiple comments "App key 5, with value, custom table and placeholder." & "Duplicate declaration of App key 5 between ObjC and Swift,and with a comment even spanning multiple lines!"
[15:24:16]: ▸ genstrings: warning: Key "pod.key5" used with multiple comments "Duplicate declaration of Pod key 5 between ObjC and Swift,and with a comment even spanning multiple lines!" & "Pod key 5, with value, custom table and placeholder."
👆👆👆👆
    logs warnings about both duplicate values and duplicate comments if not in quiet mode
```

Granted, that's only a minor annoyance, but it can make for noisy test outputs at times.

Reply:
Yeah I'm balanced about this.

On one end I totally agree that it's better to test this in isolation.
On the other end, I thought about creating a helper but then realized it felt kinda pointless to have a brand new helper file just for one method with 4 lines of code.

And also, it might actually be interesting to test the action itself, especially to test those `UI.message` outputs.
For example, I just realized that there was an unbalance in the `:quiet` parameter, as it both acts on the `-q` flag on `genstrings` (which either logs only multiple values, or both values and comments, but never deactivate all warnings) and on the `UI.command_output`. So `quiet: true` will set `-q` but also not print anything in stdout (which kinda renders the use of `-q` pointless)… Something I'd have caught with a dedicated test.
(I'll probably add a commit to change the code to provide a `log_level:{0,1,2}` instead of `quiet:{true,false}`)

Reply:
Ok so with the latest commits I pushed, I went with:
 - Switching from `Open3.capture2e` to `sh` – which is the more idiomatic way `fastlane` suggest to call shell commands anyway, and has the benefit to handle the output verbosity for us, etc
 - Which is why I kept the `:quiet` parameter as-is after all (and not `:log_level`) as it now only affects the `-q` more of `genstrings`, while the output verbosity is managed by fastlane itself.
 - I decided to still keep testing the action itself (as opposed to separate the logic in a dedicated object. Even if I moved the code into a dedicated object, it would still need to call `sh` to invoke the `genstrings` command anyway, which would try printing stuff to stdout anyway even if invoked from an helper object, so that wouldn't help with that

I might still be swayed into moving the code from the action into a dedicated helper object now that the code has changed (and that added helper methods `glob_pattern` and `files_matching` in the action) if you really feel strongly about it… but since that would not change the fact that `UI.command_output` will still be called internally by `sh` (and the "Logging disabled" warnings will still appear), as the test still has to actually call `genstrings` anyway. So given that, I feel like it might not be worth creating a helper in `helpers/` and having to switch back and forth between the 2 files while maintaining this code, just for so few lines of code, as opposed to keeping it as helper methods in the Action. WDYT?

Reply:
I agree with your point that this code is quite minimal and therefore the maintenance overhead of having the `gentstrings`-wrapping code extracted in its own object would be more than the value added by isolated testing. 👍